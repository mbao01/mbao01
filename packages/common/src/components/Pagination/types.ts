import { VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";

export type PaginationProps = React.ComponentProps<"nav">;

export type PaginationContentProps = React.ComponentProps<"ul">;

export type PaginationEllipsisProps = React.ComponentProps<"span">;

export type PaginationItemProps = React.ComponentProps<"li">;

export type PaginationLinkProps = React.ComponentProps<"a"> &
  VariantProps<typeof getButtonClasses> & {
    isActive?: boolean;
  };

export type PaginationPreviousProps = PaginationLinkProps;

export type PaginationNextProps = PaginationLinkProps;
