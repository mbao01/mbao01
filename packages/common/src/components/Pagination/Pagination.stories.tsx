import type { Meta, StoryObj } from "@storybook/react";
import { Pagination } from "./Pagination";

const withPagination = () => {
  return (
    <Pagination>
      <Pagination.Content>
        <Pagination.Item>
          <Pagination.Previous href="#" />
        </Pagination.Item>
        <Pagination.Item>
          <Pagination.Link href="#">1</Pagination.Link>
        </Pagination.Item>
        <Pagination.Item>
          <Pagination.Link href="#" isActive>
            2
          </Pagination.Link>
        </Pagination.Item>
        <Pagination.Item>
          <Pagination.Link href="#">3</Pagination.Link>
        </Pagination.Item>
        <Pagination.Item>
          <Pagination.Ellipsis />
        </Pagination.Item>
        <Pagination.Item>
          <Pagination.Next href="#" />
        </Pagination.Item>
      </Pagination.Content>
    </Pagination>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Pagination,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withPagination],
} satisfies Meta<typeof Pagination>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};
