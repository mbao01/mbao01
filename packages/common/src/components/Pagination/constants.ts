import { cva } from "../../libs";

export const getPaginationClasses = cva("flex justify-center");

export const getPaginationContentClasses = cva(
  "flex flex-row items-center gap-1"
);

export const getPaginationEllipsisClasses = cva(
  "flex h-9 w-9 items-center justify-center"
);

export const getPaginationNextClasses = cva("gap-1 px-3");

export const getPaginationPreviousClasses = cva("gap-1 px-3");
