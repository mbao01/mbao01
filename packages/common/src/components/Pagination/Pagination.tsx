import { forwardRef } from "react";
import {
  ChevronLeftIcon,
  ChevronRightIcon,
  DotsHorizontalIcon,
} from "@radix-ui/react-icons";
import {
  PaginationContentProps,
  PaginationEllipsisProps,
  PaginationItemProps,
  PaginationLinkProps,
  PaginationNextProps,
  PaginationPreviousProps,
  type PaginationProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getPaginationClasses,
  getPaginationContentClasses,
  getPaginationEllipsisClasses,
  getPaginationNextClasses,
  getPaginationPreviousClasses,
} from "./constants";

import { getButtonClasses } from "../Button/constants";

const Pagination = ({ className, ...props }: PaginationProps) => (
  <nav
    role="navigation"
    aria-label="pagination"
    className={cn(getPaginationClasses(), className)}
    {...props}
  />
);
Pagination.displayName = "Pagination";

const PaginationContent = forwardRef<HTMLUListElement, PaginationContentProps>(
  ({ className, ...props }, ref) => (
    <ul
      ref={ref}
      className={cn(getPaginationContentClasses(), className)}
      {...props}
    />
  )
);
PaginationContent.displayName = "PaginationContent";

const PaginationItem = forwardRef<HTMLLIElement, PaginationItemProps>(
  (props, ref) => <li ref={ref} {...props} />
);
PaginationItem.displayName = "PaginationItem";

const PaginationLink = ({
  className,
  size,
  wide,
  outline,
  variant = "ghost",
  isActive,
  ...props
}: PaginationLinkProps) => (
  <a
    aria-current={isActive ? "page" : undefined}
    className={cn(
      getButtonClasses({
        size,
        wide,
        outline: outline ?? isActive,
        variant,
      }),
      className
    )}
    {...props}
  />
);
PaginationLink.displayName = "PaginationLink";

const PaginationPrevious = ({
  className,
  children,
  ...props
}: PaginationPreviousProps) => (
  <PaginationLink
    aria-label="Go to previous page"
    className={cn(getPaginationPreviousClasses(), className)}
    {...props}
  >
    <ChevronLeftIcon className="h-4 w-4" />
    {children && <span>{children}</span>}
  </PaginationLink>
);
PaginationPrevious.displayName = "PaginationPrevious";

const PaginationNext = ({
  className,
  children,
  ...props
}: PaginationNextProps) => (
  <PaginationLink
    aria-label="Go to next page"
    className={cn(getPaginationNextClasses(), className)}
    {...props}
  >
    {children && <span>{children}</span>}
    <ChevronRightIcon className="h-4 w-4" />
  </PaginationLink>
);
PaginationNext.displayName = "PaginationNext";

const PaginationEllipsis = ({
  className,
  ...props
}: PaginationEllipsisProps) => (
  <span
    aria-hidden
    className={cn(getPaginationEllipsisClasses(), className)}
    {...props}
  >
    <DotsHorizontalIcon className="h-4 w-4" />
    <span className="sr-only">More pages</span>
  </span>
);
PaginationEllipsis.displayName = "PaginationEllipsis";

Pagination.Content = PaginationContent;
Pagination.Link = PaginationLink;
Pagination.Item = PaginationItem;
Pagination.Previous = PaginationPrevious;
Pagination.Next = PaginationNext;
Pagination.Ellipsis = PaginationEllipsis;

export { Pagination };
