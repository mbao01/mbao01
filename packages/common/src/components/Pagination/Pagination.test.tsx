import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { Pagination } from "./Pagination";

describe("Pagination", () => {
  const renderPagination = () =>
    render(
      <Pagination>
        <Pagination.Content>
          <Pagination.Item>
            <Pagination.Previous href="#" />
          </Pagination.Item>
          <Pagination.Item>
            <Pagination.Link href="#">1</Pagination.Link>
          </Pagination.Item>
          <Pagination.Item>
            <Pagination.Link href="#" isActive>
              2
            </Pagination.Link>
          </Pagination.Item>
          <Pagination.Item>
            <Pagination.Link href="#">3</Pagination.Link>
          </Pagination.Item>
          <Pagination.Item>
            <Pagination.Ellipsis />
          </Pagination.Item>
          <Pagination.Item>
            <Pagination.Next href="#" />
          </Pagination.Item>
        </Pagination.Content>
      </Pagination>
    );

  it("renders pagination buttons", () => {
    const { asFragment } = renderPagination();

    expect(
      screen.getByRole("link", { name: "Go to previous page" })
    ).toBeEnabled();
    expect(screen.getByRole("link", { name: "Go to next page" })).toBeEnabled();

    ["1", "2", "3"].forEach((page) => {
      expect(screen.getByRole("link", { name: page })).toBeEnabled();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
