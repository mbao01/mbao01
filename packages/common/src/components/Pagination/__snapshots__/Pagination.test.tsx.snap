// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders pagination buttons 1`] = `
<DocumentFragment>
  <nav
    aria-label="pagination"
    class="flex justify-center"
    role="navigation"
  >
    <ul
      class="flex flex-row items-center gap-1"
    >
      <li>
        <a
          aria-label="Go to previous page"
          class="btn btn-ghost min-h-fit h-10 rounded-md gap-1 px-3"
          href="#"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-left h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15 18-6-6 6-6"
            />
          </svg>
        </a>
      </li>
      <li>
        <a
          class="btn btn-ghost min-h-fit h-10 rounded-md"
          href="#"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-current="page"
          class="btn btn-ghost btn-outline min-h-fit h-10 rounded-md"
          href="#"
        >
          2
        </a>
      </li>
      <li>
        <a
          class="btn btn-ghost min-h-fit h-10 rounded-md"
          href="#"
        >
          3
        </a>
      </li>
      <li>
        <span
          aria-hidden="true"
          class="flex h-9 w-9 items-center justify-center"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-ellipsis h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="1"
            />
            <circle
              cx="19"
              cy="12"
              r="1"
            />
            <circle
              cx="5"
              cy="12"
              r="1"
            />
          </svg>
          <span
            class="sr-only"
          >
            More pages
          </span>
        </span>
      </li>
      <li>
        <a
          aria-label="Go to next page"
          class="btn btn-ghost min-h-fit h-10 rounded-md gap-1 px-3"
          href="#"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-right h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
